### PR TITLE
#284 improve Emitter 

### PR DESCRIPTION
**Pull request by Emily-Jiang**
Signed-off-by: Emily Jiang <emijiang6@googlemail.com>

### CODE REVIEW

**Comment by Emily-Jiang** [comment]
@cescoffier can you take a look to see why the tests are failing? What is the completion criteria for the CompletionStage to complete?

File: smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/extension/EmitterImpl.java [hunk 1]
```diff
@@ -67,18 +73,18 @@ boolean isConnected() {
...
-    public synchronized Emitter<T> send(T msg) {
+    public synchronized CompletionStage<Void> send(T msg) {
         if (msg == null) {
             throw new IllegalArgumentException("`null` is not a valid value");
         }
         FlowableEmitter<Message<? extends T>> emitter = verify();
         if (msg instanceof Message) {
             //noinspection unchecked
-            emitter.onNext((Message) msg);
+            return CompletableFuture.runAsync(() -> emitter.onNext((Message) msg));
```
Comment (cescoffier) [inline smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/extension/EmitterImpl.java]: That's not correct.

You need something like (be aware this is pseudo-code):

```
CompletableFuture future = new CompletableFuture();
if (msg instanceof Message) {
    // here things need to be discussed
    // if the user passes a Message, he can attach the acknowledgment logic
   // however we cannot really do anything. We would need a way to customize the Supplier
} else {
   emitter.onNext(Message.of(msg, () -> future.complete(null));
}
```

File: smallrye-reactive-messaging-provider/src/test/java/io/smallrye/reactive/messaging/inject/EmitterInjectionTest.java [hunk 2]
```diff
@@ -268,13 +274,15 @@ public boolean isCaught() {
         }
 
         public void run() {
-            emitter.send("a").send("b");
+            emitter.send("a");
```
Comment (cescoffier) [inline smallrye-reactive-messaging-provider/src/test/java/io/smallrye/reactive/messaging/inject/EmitterInjectionTest.java]: The tests need to check that the returned CompletionStage is completed when the message is processed.

File: smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/annotations/Emitter.java [hunk 3]
```diff
@@ -11,27 +13,28 @@
...
 public interface Emitter<T> {
 
     /**
      * Sends a payload or a message to the stream.
      *
      * @param msg the <em>thing</em> to send, must not be {@code null}
-     * @return the current emitter
-     * @throws IllegalStateException if the stream does not have any pending requests, or if the stream has been
-     *         cancelled or terminated.
+     * @return the {@code CompletionStage}
```
Comment (cescoffier) [inline smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/annotations/Emitter.java]: it needs to explain when the CS is completed.

File: smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/annotations/OnOverflow.java [hunk 4]
```diff
@@ -35,10 +35,17 @@
...
-         * buffer. If the buffer is full, an error is propagated.
+         * This creates a buffer with the size specified by {@link #bufferSize()} if present.
+         * Otherwise, the size will be the value of the config property
+         * <strong>mp.messaging.emitter.defult-buffer-size</strong>.
+         * If the buffer is full, an error is propagated.
          */
         BUFFER,
+        /**
+         * Buffers <strong>all</strong> values until the downstream consumes it.
+         * This creates an unbound buffer. If the buffer is full, an error is propagated.
```
Comment (cescoffier) [inline smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/annotations/OnOverflow.java]: _If the buffer is full_... in this case if the buffer is full, you are running out of memory :-)

**Comment by cescoffier** [comment]
There is a design issue when the user passes a `Message`, as we cannot return a `CompletionStage` completed when the message is acknowledged.

One way would be to have 2 methods:

* `void send(Message<? extends T> message);` // can also return `this`
* `CompletionStage<Void> send(T payload);` 

We need to check how practical is it.

**Comment by cescoffier** [comment]
For the record, the attached patch should fix the test failures and fix the generic signature.
[emitter.patch.txt](https://github.com/smallrye/smallrye-reactive-messaging/files/3868391/emitter.patch.txt)

**Comment by cescoffier** [comment]
Remaining tasks:

- [ ] Test with a custom message implementation
- [ ] Test that the ack is correctly propagated to the returned CompletionStage.

File: smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/annotations/Emitter.java (hunk 3 above)
Comment (Emily-Jiang) [reply smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/annotations/Emitter.java]: done

File: smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/annotations/OnOverflow.java (hunk 4 above)
Comment (Emily-Jiang) [reply smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/annotations/OnOverflow.java]: done

File: smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/extension/EmitterImpl.java (hunk 1 above)
Comment (Emily-Jiang) [reply smallrye-reactive-messaging-provider/src/main/java/io/smallrye/reactive/messaging/extension/EmitterImpl.java]: updated

File: smallrye-reactive-messaging-provider/src/test/java/io/smallrye/reactive/messaging/inject/EmitterInjectionTest.java (hunk 2 above)
Comment (Emily-Jiang) [reply smallrye-reactive-messaging-provider/src/test/java/io/smallrye/reactive/messaging/inject/EmitterInjectionTest.java]: added more tests

**Comment by Emily-Jiang** [comment]
@cescoffier please check whether you have more comments.

**Comment by cescoffier** [comment]
The Kafka tests are failing because they use an emitter.
Apply the following diff:

```
         @Inject
         @Channel("temperature-values")
-        private Emitter<KafkaMessage<String, String>> emitter;
+        private Emitter<String> emitter;
```

Also, it would be great to rebase and cleanup the history a bit.

**Comment by cescoffier** [comment]
We need more tests:
1. Test emitter with messages
2. Test that the acknowledgment is propagated correctly. This set of tests is very important because it's the core of this PR: the completion stage returned by the emitter but be completed when the message is acknowledged.

**Comment by cescoffier** [comment]
Also, this PR is a breaking change on the API, so, after discussing with @kenfinnigan we would need to:

* keep the previous behavior for the Emitter class contained in the provider module (the old one, already deprecated)
* use the new API for the Emitter class from the `api` package.

**Comment by Emily-Jiang** [comment]
I have added the tests to check the acknowledgment was propagated correctly though.

File: smallrye-reactive-messaging-provider/src/test/java/io/smallrye/reactive/messaging/inject/EmitterInjectionTest.java [hunk 5]
```diff
@@ -137,21 +167,24 @@ public void testWithTwoEmittersConnectedToOneProcessor() {
...
+        public void consume(final String s) {
             list.add(s);
         }
     }
 
     @ApplicationScoped
     public static class MyBeanEmittingPayloadsUsingStream {
         @Inject
-        @Stream("foo")
+        @Channel("foo")
```
Comment (cescoffier) [inline smallrye-reactive-messaging-provider/src/test/java/io/smallrye/reactive/messaging/inject/EmitterInjectionTest.java]: This test is about testing the legacy Stream annotation, do don't update to @Channel.

File: smallrye-reactive-messaging-provider/src/test/java/io/smallrye/reactive/messaging/inject/EmitterInjectionTest.java [hunk 6]
```diff
@@ -220,16 +380,16 @@ public void run() {
...
-        public void consume(String s) {
+        public void consume(final String s) {
             list.add(s);
         }
     }
 
     public static class BeanWithMissingStream {
         @Inject
-        @Stream("missing")
+        @Channel("missing")
```
Comment (cescoffier) [inline smallrye-reactive-messaging-provider/src/test/java/io/smallrye/reactive/messaging/inject/EmitterInjectionTest.java]: This test is testing the behavior of the legacy @Stream annotation, so don't replace it.

Comment (Emily-Jiang) [reply smallrye-reactive-messaging-provider/src/test/java/io/smallrye/reactive/messaging/inject/EmitterInjectionTest.java]: done

File: smallrye-reactive-messaging-provider/src/test/java/io/smallrye/reactive/messaging/inject/EmitterInjectionTest.java (hunk 5 above)
Comment (Emily-Jiang) [reply smallrye-reactive-messaging-provider/src/test/java/io/smallrye/reactive/messaging/inject/EmitterInjectionTest.java]: done

**Comment by Emily-Jiang** [comment]
@cescoffier I tried to add tests to test the deprecated Emitter. However, all plugging was removed by the package refractoring and the new ground needs to be laid out. I think if you want to rewire the old api, all of the scaffolding needs to be done again. This should be done under a different issue. By the way, I've addressed all of your comments except this one.

**Comment by cescoffier** [comment]
Starting working on it to include the PR _soonish_.

**Comment by cescoffier** [comment]
This PR is superseded by #339

**Comment by Emily-Jiang** [comment]
Close this one because #339 superseded this.

**Pull request closed without merging**